pkg/gadget-collection/gadgets/traceloop/gadget.go

```diff
@@ -60,130 +77,423 @@ some differences:
+		}
+
+		trace.Status.OperationError = fmt.Sprintf("%q is not a valid ID for this trace, valid IDs are: %v", containerID, strings.Join(ids, ","))
+
+		return
 	}
 
-	err := t.cmd.Process.Signal(syscall.SIGINT)
+	traceUnique.Lock()
+	events, err := traceUnique.tracer.Read(containerID)
+	traceUnique.Unlock()
 	if err != nil {
-		return fmt.Errorf("failed to send SIGINT to process: %w", err)
+		trace.Status.OperationError = fmt.Sprintf("Failed to read perf buffer: %s", err)
+
+		return
+	}
+
+	traceName := gadgets.TraceName(trace.ObjectMeta.Namespace, trace.ObjectMeta.Name)
+	r, err := json.Marshal(events)
+	if err != nil {
+		log.Warnf("Gadget %s: error marshalling event: %s", trace.Spec.Gadget, err)
+		return
+	}
+	// HACK Traceloop is really particular as it cannot use Status output because
+	// the size is limited.
+	// Also, if we send each event as a line, we will only be able to get the last
+	// 100 (or 250) events from the CLI due to this code:
+	// https://github.com/inspektor-gadget/inspektor-gadget/blob/9c7b6a126d82b54262ffdc5709d7c92480002830/pkg/gadgettracermanager/stream/stream.go#L24
+	// https://github.com/inspektor-gadget/inspektor-gadget/blob/9c7b6a126d82b54262ffdc5709d7c92480002830/pkg/gadgettracermanager/stream/stream.go#L95-L100
+	// To overcome this limitation, we just send all events as one big line.
+	// Then, the CLI receives this big line, parses it and prints each event.
+	// A proper solution would be to develop a specific "output" (neither status
+	// nor stream) for traceloop, this is let as TODO.
+	t.helpers.PublishEvent(traceName, string(r))
+
+	trace.Status.State = gadgetv1alpha1.TraceStateCompleted
+}
+
+func (t *Trace) Delete(trace *gadgetv1alpha1.Trace) {
+	if trace.Spec.OutputMode != gadgetv1alpha1.TraceOutputModeStatus {
+		trace.Status.OperationError = fmt.Sprintf("\"delete\" operation can only be used with %q trace while %q was given", gadgetv1alpha1.TraceOutputModeStatus, trace.Spec.OutputMode)
+
+		return
 	}
 
-	timeout := time.After(2 * time.Second)
+	containerID := trace.Spec.Parameters["containerID"]
+	mntNsID, ok := t.containerIDs[containerID]
+	if !ok {
+		ids := make([]string, len(t.containerIDs))
+		i := 0
+		for id := range t.containerIDs {
+			ids[i] = id
+			i++
+		}
 
-	done := make(chan struct{})
-	go func() {
-		t.cmd.Process.Wait()
-		done <- struct{}{}
-	}()
+		trace.Status.OperationError = fmt.Sprintf("%q is not a valid ID for this trace, valid IDs are: %v", containerID, strings.Join(ids, ","))
 
-	select {
-	case <-timeout:
-		t.cmd.Process.Kill()
-	case <-done:
+		return
 	}
 
-	t.started = false
+	traceUnique.Lock()
+	if traceUnique.tracer == nil {
+		traceUnique.Unlock()
+
+		trace.Status.OperationError = "Traceloop tracer is nil"
+
+		return
+	}
+
+	// First, we need to detach the perf buffer.
+	err := traceUnique.tracer.Detach(mntNsID)
+	if err != nil {
+		traceUnique.Unlock()
+
+		trace.Status.OperationError = fmt.Sprintf("Failed to remove perf buffer from map: %s", err)
+
+		return
```

Comment:
The error should be ignored: if the container has terminated, then it is already detached.

```
$ kubectl gadget traceloop delete 19869dcf2105eabc7a
Error: failed to run gadget on node "minikube": Failed to remove perf buffer from map: error removing perf buffer from map with mntnsID 4026533226
```

I cannot delete this trace because of this problem.

Actually, since mntns id can be recycled, traceloop needs to keep track of containers that have been detached, so it would only detach those who have not been detached.